File: python/fuzz/targets/jws.py
"""The StoreKit 2 path: compact-JWS split, strict canonical base64url, the
JSON header and payload, the ``x5c`` certificates, the chain, the ES256
signature, then each entry point's claim checks.

Same invariants as the Go port's ``FuzzVerifyTransaction``: nothing escapes
but a ``VerificationError``, and a JWS that ``verify_raw`` accepts under the
fixture root must be refused under Apple's roots, or the anchors are not what
decided it.

The header and payload are attacker-supplied JSON, so this target is the one
that reaches the library's claim reads with values of the wrong *type* and
the wrong *magnitude* — the two shapes a hand-written `.get()` chain tends to
miss.
"""

from harness import (
    APPLE_JWS_VERIFIER,
    JWS_VERIFIER,
    InvariantViolation,
    as_text,
    require_verification_error,
    run,
)

_CALLS = (
    ("verify_transaction", JWS_VERIFIER.verify_transaction),
    ("verify_app_transaction", JWS_VERIFIER.verify_app_transaction),
    ("verify_raw", JWS_VERIFIER.verify_raw),
)


def one_input(data: bytes) -> None:
    jws = as_text(data)
    if jws is None:
        return
    accepted_by_fixture_root = False
    for name, call in _CALLS:
        try:
            call(jws)
            accepted_by_fixture_root = accepted_by_fixture_root or name == "verify_raw"
        except Exception as error:
            require_verification_error(error, name)
    if not accepted_by_fixture_root:
        return
    try:
        APPLE_JWS_VERIFIER.verify_raw(jws)
    except Exception as error:
        require_verification_error(error, "verify_raw against Apple's roots")
        return
    raise InvariantViolation(
        "this input verifies against Apple's roots too, so the anchors are not being enforced"
    )


if __name__ == "__main__":
    run(one_input)

File: dotnet/fuzz/Targets/Jws.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using ApplePurchaseReceiptVerifier.Jws;

namespace ApplePurchaseReceiptVerifier.Fuzz.Targets
{
    /// <summary>
    /// The StoreKit 2 path: compact-JWS split, strict base64url, JSON header
    /// and payload, the <c>x5c</c> certificates, the chain, the ES256
    /// signature, then the claim checks of all three public entry points.
    /// </summary>
    /// <remarks>
    /// Same invariants as the Go port's <c>FuzzVerifyTransaction</c>: nothing
    /// but a <see cref="VerificationException"/> escapes any of the three, and
    /// a JWS that <c>VerifyRaw</c> accepts under the generated fixture root
    /// must be refused under Apple's real JWS roots — otherwise the anchors
    /// are not what decided it.
    /// </remarks>
    internal sealed class Jws : IDisposable
    {
        private readonly JwsVerifier _fixture;
        private readonly JwsVerifier _unrelated;
        private readonly X509Certificate2 _root;

        internal Jws()
        {
            _root = Fixtures.JwsRoot();
            AppleEnvironment[] environments =
            {
                AppleEnvironment.Production,
                AppleEnvironment.Sandbox,
                AppleEnvironment.Xcode,
                AppleEnvironment.LocalTesting,
            };
            _fixture = new JwsVerifier(
                new[] { _root }, "com.example.app", environments, appAppleId: 1234567890);
            _unrelated = new JwsVerifier(
                AppleRootCertificates.JwsRoots(), "com.example.app", environments, appAppleId: 1234567890);
        }

        internal void Run(ReadOnlySpan<byte> data)
        {
            string jws;
            try
            {
                jws = new UTF8Encoding(false, true).GetString(data.ToArray());
            }
            catch (DecoderFallbackException)
            {
                return;
            }

            try
            {
                _fixture.VerifyTransaction(jws);
            }
            catch (Exception e)
            {
                Invariant.Contained("VerifyTransaction", e);
            }

            try
            {
                _fixture.VerifyAppTransaction(jws);
            }
            catch (Exception e)
            {
                Invariant.Contained("VerifyAppTransaction", e);
            }

            try
            {
                _fixture.VerifyRaw(jws);
            }
            catch (Exception e)
            {
                Invariant.Contained("VerifyRaw", e);
                return;
            }

            try
            {
                _unrelated.VerifyRaw(jws);
            }
            catch (VerificationException)
            {
                return;
            }

            throw new InvariantException(
                "this input verifies against Apple's roots too, "
                + "so the anchors are not being enforced");
        }

        public void Dispose()
        {
            _fixture.Dispose();
            _unrelated.Dispose();
            _root.Dispose();
        }
    }
}
